source/extensions/filters/http/grpc_field_extraction/filter_config.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include <cstdlib>
+#include <memory>
+#include <optional>
+#include <string>
+#include <utility>
+
+#include "envoy/extensions/filters/http/grpc_field_extraction/v3/config.pb.h"
+#include "envoy/extensions/filters/http/grpc_field_extraction/v3/config.pb.validate.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/server/filter_config.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/grpc/common.h"
+#include "source/common/protobuf/protobuf.h"
+#include "source/extensions/filters/http/grpc_field_extraction/extractor.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/status/status.h"
+#include "absl/strings/string_view.h"
+#include "grpc_transcoding/type_helper.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace GrpcFieldExtraction {
+
+class FilterConfig : public Envoy::Logger::Loggable<Envoy::Logger::Id::filter> {
+public:
+  explicit FilterConfig(
+      const envoy::extensions::filters::http::grpc_field_extraction::v3::GrpcFieldExtractionConfig&
+          proto_config,
+      std::unique_ptr<ExtractorFactory> extractor_factory, Api::Api& api);
+
+  const Extractor* FindExtractor(absl::string_view proto_path) const;
+
+private:
+  void initDescriptorPool(Api::Api&);
+  const envoy::extensions::filters::http::grpc_field_extraction::v3::GrpcFieldExtractionConfig
+      proto_config_;
+  absl::flat_hash_map<std::string, std::unique_ptr<Extractor>> proto_path_to_extractor_;
```

Comment:
This state is accessed from multiple threads and needs to be immutable.  Can you please make all members `const`  or `const std::unique_ptr<const T>` to allow compiler to validate immutability? Otherwise it is hard to reason about its correctness, especially for complex objects like `DescriptorPool`

Reply:
IMHO, it is overkill after we make the FilterConfig as const object and all public methods as const.

Reply:
Agreed with @TAOXUY. IIUC this object is created once via config thread. All worker threads have read-only access on this object so it is thread-safe by definition

@TAOXUY maybe you can add a class comment stating this is is thread-safe

Reply:
Updated the comment.

Reply:
It is not enough to use `std::shared_ptr<const FilterConfig>` if you use std::unqiue_ptr (or similar classes). Note that constness of the FilterConfig only males the smart pointer wrapper const (i.e. you can not call .reset() on it). However it does not make the object it points to const. For example you will still be able to call mutable methods through `type_finder_`.

The same applies to the proto_path_to_extractor_. It needs to be const itself and the `std::unique_ptr<const Extractor>` to be fully immutable.

Reply:
DONE.